### DISCUSSION TITLE
CrazyFlie Flowdeck V2 for Location Tracking

### DISCUSSION THREAD

**Discussion by ethan-sd** [bitcraze/discussions#1546] [Q&A]
As a small project, we are trying to use a Flowdeck connected to a Crazyflie to track location from movement within a room. 

Upon testing, the cfclient either detects no motion (x=0, y=0), or has the x and y values increasing linearly, proportional with one another.
![image](https://github.com/user-attachments/assets/b3acb725-a301-4666-9e54-d384c684f710)

In some test cases, flowdeck was able to record a dirty dataset, recording as it was brought around a room. 
![image](https://github.com/user-attachments/assets/e09a850e-45bd-4d68-a5b9-052afb4ea9dd)

 Would like to ask for advice on how to implement this system, if flowdeck is the right choice for this or not :)

**Comment by tobbeanton** [comment]
The flow deck is going to have errors which is dependent on the surface it tracks. During good tracking conditions it shouldn't be more then centimeters per meter but as the tracking is relative it is are going to add up. Perhaps [this](https://www.bitcraze.io/2023/11/go-with-the-flow-relative-positioning-with-the-flow-deck/) post can help a bit.

**Comment by ethan-sd** [comment]
Hi Tobias!

Thank you for helping us address our concerns.

Would you have any advice for the process of accounting for yaw?

Our specific application requires yaw changes to be recorded and
characterized to map the pathing of the drone accurately. Currently, the
yaw seems to be limited by the rate of rotation. When it turns 90 degrees,
the yaw value seems to be undervalued.

Hoping to hear from you soon, if you would be free to meet or would prefer
a different mode of communication let me know!

Best,
Ethan

**Reply by tobbeanton** [reply]
Unfortunately yaw is not tracked by the optical flow sensor, only by the gyro, which also has drift. If you want some form of absolute tracking the flow is not going to be good enough. You would need some added/other positioning system.

**Comment by ethan-sd** [comment]
Thank you once again Tobias for the help!

Would you be able to help us characterize the drift from the gyro onboard?

Yaw seems to change more for us as the drone rotates faster, with the yaw
going undetected with slow rotations.

Additionally, from our testing, the distance measurements would shift more
drastically when the drone moves much faster.

Hoping for your assistance on this!

Best,
Ethan

**Reply by knmcguire** [reply]
To measure the gyrodrift is not exactly easy, since it is different for every system. Also we already do some sorts of gyro calibration at the beginning of startup of the crazyflie.

I think that you'll need to have an separate positioning system like motion capture to evaluate the gyrodrift during flight. Lighthouse positioning also does absolute yaw estimation so you could use that as well, but you'll have to make sure that the gyro measurements don't influence the state estimate from the filter, so that will be more tricker to really do that properly

I'm afraid that the gyrodrift is something that the current extended kalman filter doesn't really have implemented, but that is from the top of my head. I advise you to perhaps look at the code and see if this is something that you can add. Here is the documentation for it: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/state_estimators/#extended-kalman-filter

Also your first issue that you were having, about not getting any good positioning estimate at all from the flowdeck, make sure to read this blogpost for tips: https://www.bitcraze.io/2023/11/go-with-the-flow-relative-positioning-with-the-flow-deck/
